Applications/Spire/Source/Ui/DropDownBox.cpp

```diff
@@ -0,0 +1,194 @@
+  container_layout->setContentsMargins({});
+  container_layout->setSpacing(0);
+  m_text_box = make_label("");
+  m_text_box->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+  m_text_box->setAttribute(Qt::WA_TransparentForMouseEvents);
+  container_layout->addWidget(m_text_box);
+  auto icon_layout = new QVBoxLayout();
+  icon_layout->setContentsMargins({});
+  icon_layout->setSpacing(0);
+  icon_layout->addStretch();
+  auto drop_down_icon = new Icon(
+    imageFromSvg(":/Icons/dropdown-arrow.svg", scale(6, 4)));
+  drop_down_icon->setFixedSize(scale(6, 4));
+  drop_down_icon->setFocusPolicy(Qt::NoFocus);
+  icon_layout->addWidget(drop_down_icon);
+  icon_layout->addStretch();
+  container_layout->addLayout(icon_layout);
+  m_input_box = make_input_box(container);
+  m_button = new Button(m_input_box);
+  auto layout = new QHBoxLayout(this);
+  layout->setContentsMargins({});
+  layout->addWidget(m_button);
+  proxy_style(*this, *m_input_box);
+  set_style(*this, DEFAULT_STYLE());
+  setFocusProxy(m_button);
+  m_drop_down_list = new DropDownList(*m_list_view, this);
+  auto list_model = m_list_view->get_list_model();
+  for(auto i = 0; i < list_model->get_size(); ++i) {
+    if(auto item = m_list_view->get_list_item(i)) {
+      item->setFocusPolicy(Qt::NoFocus);
+    }
+  }
+  m_button->connect_clicked_signal([=] { on_click(); });
+  m_list_view_current_connection =
+    m_list_view->get_current_model()->connect_current_signal(
+      [=] (auto& current) { on_list_view_current(current); });
+  m_list_view_submit_connection = m_list_view->connect_submit_signal(
+    [=] (auto& submission) { on_list_view_submit(submission); });
+  m_button->installEventFilter(this);
+  m_drop_down_list->window()->installEventFilter(this);
+  m_list_view->installEventFilter(this);
+}
+
+bool DropDownBox::is_read_only() const {
+  return m_is_read_only;
+}
+
+void DropDownBox::set_read_only(bool is_read_only) {
+  m_is_read_only = is_read_only;
+  if(m_is_read_only) {
+    match(*this, ReadOnly());
+    match(*m_input_box, ReadOnly());
+  } else {
+    unmatch(*this, ReadOnly());
+    unmatch(*m_input_box, ReadOnly());
+  }
+}
+
+connection DropDownBox::connect_submit_signal(
+    const SubmitSignal::slot_type& slot) const {
+  return m_submit_signal.connect(slot);
+}
+
+bool DropDownBox::eventFilter(QObject* watched, QEvent* event) {
+  if(event->type() == QEvent::KeyPress) {
+    if(m_drop_down_list->isVisible()) {
+      auto key_event = static_cast<QKeyEvent*>(event);
+      switch(key_event->key()) {
+        case Qt::Key_Tab:
+        case Qt::Key_Backtab:
+          m_drop_down_list->hide();
+          QCoreApplication::sendEvent(m_button, event);
+          break;
+        case Qt::Key_Escape:
+          m_list_view->get_current_model()->set_current(m_submission_index);
+          m_drop_down_list->hide();
+          break;
+      }
+    }
+  } else if(event->type() == QEvent::Close &&
+      watched == m_drop_down_list->window()) {
```

Comment:
It might be better to group the events that are filtered by the m_drop_down_list under a common branch of these if statements, e.g.;

```
if(event->type() == ...) {
  ...
} else if(event->type() == ...) {
  ...
} else if(watched == m_drop_down_list) {
  if(event->type() == ...) {

  } else if(event->type() == ...) {

  }
}
```

This way it's clear at a glance that these events are only filtered by the m_drop_down_list, but the other events are shared by all filters.